File: leetcode/go/2976.minimum-cost-to-convert-string-i.go
func minimumCost(source string, target string, original []byte, changed []byte, cost []int) int64 {
    or := []byte(original)
    ch := []byte(changed)
    cc := [26][26]int64{}
    for i := range 26 {
        for j := range 26 {
            cc[i][j] = math.MaxInt64
        }
    }
    for i, o := range or {
        cc[o-97][ch[i]-97] = min(cc[o-97][ch[i]-97], int64(cost[i]))
    }
    for k := range 26 {
        for i := range 26 {
            if cc[i][k] < math.MaxInt64 {
                for j := range 26 {
                    if cc[k][j] < math.MaxInt64 {
                        cc[i][j] = min(cc[i][j], cc[i][k] + cc[k][j])
                    }
                }
            }
        }
    }
    sb := []byte(source)
    tb := []byte(target)
    var ans int64 = 0
    for i, s := range sb {
        if s != tb[i] {
            if cc[s-97][tb[i]-97] == math.MaxInt64 {
                return -1
            }
            ans += cc[s-97][tb[i]-97]
        }
    }
    return ans
}

func min(a, b int64) int64 {
    if a < b {
        return a
    }
    return b
}

File: leetcode/rust-leetcode/src/2976.minimum-cost-to-convert-string-i.rs
impl Solution {
    pub fn minimum_cost(source: String, target: String, original: Vec<char>, changed: Vec<char>, cost: Vec<i32>) -> i64 {
        let mut cc = vec![vec![u64::MAX; 26]; 26];
        let (mut o, mut c) = (0, 0);
        for i in 0..original.len() {
            o = original[i] as usize - 97;
            c = changed[i] as usize - 97;
            cc[o][c] = cc[o][c].min(cost[i] as u64);
        }
        for k in 0..26 {
            for i in 0..26 {
                if cc[i][k] < u64::MAX {
                    for j in 0..26 {
                        if cc[k][j] < u64::MAX {
                            cc[i][j] = cc[i][j].min(cc[i][k] + cc[k][j]);
                        }
                    }
                }
            }
        }
        let mut ans = 0;
        for (s, t) in source.as_bytes().into_iter().zip(target.as_bytes()) {
            if s != t {
                (o, c) = ((s - b'a') as usize, (t-b'a') as usize);
                if cc[o][c] == u64::MAX {
                    return -1;
                }
                ans += cc[o][c];
            }
        }
        ans as _
    }
}

